runtime/src/util/map.rs

```diff
@@ -0,0 +1,164 @@
+use crate::builtin::HAMT_BIT_WIDTH;
+use crate::{ActorError, AsActorError, Hasher};
+use anyhow::anyhow;
+use cid::Cid;
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_hamt::{BytesKey, Error, Hamt};
+use fvm_shared::error::ExitCode;
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use std::fmt::Debug;
+
+/// Wraps a HAMT to provide a convenient map API.
+/// Any errors are returned with exit code indicating illegal state.
+/// The name is not persisted in state, but adorns any error messages.
+pub struct Map2<'bs, BS, K, V>
+where
+    BS: Blockstore,
+    K: Into<BytesKey> + Debug,
```

Comment:
Yeah, this `BytesKey` is annoying. Can we have a trait instead?

I.e.:

```rust
trait HamtKey {
    fn from_bytes(&[u8]) -> Result<Self>;
    fn to_bytes(&self) -> Result<Vec<u8>>;
}
```

Then we'd use `BytesKey` internally.

We don't _need_ a custom trait (could just use `From`/`Into`) but it would be nice to be a bit more explicit here.

Reply:
I reverted to just using `&[u8]` as the key type, no generic parameter. The trait as you suggested would still require callers to wrap whatever type they really want. from_bytes would be nice for some for_each users, but some others don't even inspect the key anyway.

Reply:
>  The trait as you suggested would still require callers to wrap whatever type they really want.

By defining a trait here, you can implement that trait for all types you care about. But I don't know how many custom key types we have (I assume it's mostly CID, tuple of CIDs, etc.).

Reply:
> you can implement that trait for all types you care about

I don't want to do that here, though. That knowledge of "all types" belongs to the callers. The types might not even be visible here. A similar problem to the actor error downcasting thing.

Reply:
Well, technically the callers can implement it except.... well, if they want to use some foreign type. So yeah, it might be a pain. But, IMO, constantly decoding/encoding keys is worse.

Reply:
Ok, done.